docs/designs/keras_callback.md

```diff
@@ -0,0 +1,301 @@
+* Case 4: Callback to perform early stopping when certain conditions are met. For example, the metrics are met after an evaluation job.
+* Case 5: Callback to export model using SavedModel after the training is completed.
+* Case 6: Callback to upload model to remote storage after the training is completed.
+
+The following, we will design how to define and implement callbacks to support those cases.
+
+## Define Callbacks for the Model in ElasticDL
+
+In the model definition file, users can add `callbacks` API to define callbacks like:
+```python
+From elasticdl.python.callbacks import (
+    PredictionOutputsProcessor,
+    LearningRateScheduler
+)
+
+def custom_model():
+    ...
+
+def callbacks():
+    prediction_output_processor = PredictionOutputsProcessor(process_fn)
+    learning_rate_scheduler = LearningRateScheduler(schedule_fn)
+    return [prediction_output_processor, learning_rate_scheduler]
+```
+
+## Initialize Callbacks and Set Callback Attributes in ElasticDL
+
+### Use a Container for Callbacks Defined in the Model
+We may define several callbacks for the model in a job. TensorFlow creates a container `CallbackList` to wrap the callbacks to conveniently call the methods in callbacks. For example:
+```python
+class CallbackList():
+    def on_batch_end(self, batch, logs=None):
+        for callback in self.callback_list:
+            callback.on_batch_end(bach, logs)
+```
+For detail, we can view the [source code](https://github.com/tensorflow/tensorflow/blob/cf7fcf164c9846502b21cebb7d3d5ccf6cb626e8/tensorflow/python/keras/callbacks.py#L189-L196
+) of `CallbackList` in TensorFlow.
+
+So, we can also use CallbackList to wrap the callbacks in ElasticDL.
+```python
+from tensorflow.python.keras.callbacks import CallbackList
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.on_batch_end(batch)
+```
+
+### Set Default Attributes for Callbacks.
+There are `set_model` and `set_params` in `tf.keras.callbacks.Callback`. `set_model` can set a model object to the attribute `model` of callback and `set_params` can set a dict object to the attribute `params` of callback.
+We also can use those methods of `CallbackList` to set `model` and `params` for the callbacks in `CallbackList`. 
+```python
+model = custom_model()
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.set_model(model)
+
+params = {
+    'batch_size': batch_size,
+    'epochs': epochs,
+    'saved_model_path': saved_model_path,
+    'checkpoint_path': checkpoint_path
+}
+callback_list.set_params(params)
+```
+
+Then, we can call `model` and `params` in the methods like:
+```python
+class CustomCallback(tf.keras.callbacks.Callback):
+    def on_train_batch_begin(self, batch, logs=None):
+        lr = self.model.optimizer.learning_rate,
+        saved_model_path = self.params.get("save_model_path")
+```
+
+
+## Implement Callbacks to Support the Cases in the Motivation
```

Comment:
For each callback, please add the description about which role (master/worker/ps) will execute this callback.